PI-3473: Netcdf saving of ancillary variables

pp-mo opened the pull request:
Saving ancillary variables, with some testcases based on CDL comparisons.

~There is some unfinished business here where the tests might be improved,~
~if and when they can use other aspects of ancillary support :~
**UPDATE:**
FIxed that.
Nearly good to go?

pp-mo commented:
Hi @stephenworsley 
Now rebased, should be GTG ! :pray:

stephenworsley commented on lib/iris/tests/test_netcdf.py:
```diff
@@ -964,6 +964,111 @@ def test_no_global_attributes(self):
+        testcube_2 = testcube_1.copy()
+        testcube_2.units = "m"
+        testcube_2.rename("alternate_data")
+        # TODO: this should not be necessary, when cube.copy() handles ancils.
```
Now that #3546 is merged, does rebasing and modifying this code work as expected?

pp-mo replied:
Well, it probably *ought* to fail then, I think, complaining that the ancillary to be added is identical to one already there.
I will investigate  this ...

stephenworsley commented on lib/iris/tests/test_netcdf.py:
```diff
@@ -964,6 +964,111 @@ def test_no_global_attributes(self):
+
+        ancil2 = ancil1.copy()
+        ancil2.data[0, 0] += 1.0
+        # TODO: needs modifying, when cube.copy() handles ancils.
```
As above, this should be possible to address now #3546 is merged.

pp-mo replied:
:+1:

pp-mo replied:
Note: this actually did *not* need code changing.
The cube is copied before an ancil is attached, so no collision between the two different versions occurs when adding to the second cube.
The two are stored separately (with disambiguated var_names), _because_ they contain different data.  
All good :+1: !

pp-mo commented on lib/iris/fileformats/netcdf.py:
```diff
@@ -1093,14 +1093,18 @@ def write(
         # variable to them
         self._add_cell_measures(cube, cf_var_cube, dimension_names)
 
+        # Add the ancillary_variables variables and associate the data variable
```
"Variable names" is really _wrong_, I think.
What all these "_add" functions do is to is create variables in the datafile.

The line above just says "Add coordinate variables", which I think is the right way to describe it.

stephenworsley replied:
Ah, okay I think I mixed up the red and green lines above here.

pp-mo commented on lib/iris/fileformats/netcdf.py:
```diff
@@ -1714,26 +1760,79 @@ def _create_cf_cell_measure_variable(
+            cube, dimension_names, ancillary_variable
+        )
+
+    # TODO: this was renamed from "_create_cf_variable" .
```
Not that I can think of.
Good spot! 😀

pp-mo commented:
> it probably ought to fail then, I think, complaining that the ancillary to be added is identical to one already there.

Rebased, and it did fail as predicted :+1:
Now fixed.
